### DISCUSSION TITLE
Invariant Union?

### DISCUSSION THREAD

**Discussion by treykeown** [python/typing#1574] [Q&A] [answered]
Is it possible to create a `TypeAlias` of a union, but where one of the specified types in a union must be matched exactly (invariant)?
```py
Special: TypeAlias = Foo | Bar
```
Subclasses of `Foo` or `Bar` would not satisfy `Special`.

I believe I can create a `TypeVar` that achieves something similar:
```py
T = TypeVar("T", Foo, Bar)
```
but my concern is that I'd be using a type variable where a "normal" annotation works best.

**Comment by hauntsaninja** [comment]
Probably not, could you talk more about your use case?

Note value constrained TypeVars may not work the way you expect them to: https://mypy-play.net/?mypy=latest&python=3.12&gist=6768354f1bab66850422a86eb5723a1f (the TypeVar is prevented from solving to `SubFoo`, not prevented from solving). Also note invariant generics don't prevent contained types from being subclasses either: https://mypy-play.net/?mypy=latest&python=3.12&gist=c29b3396aabe1ef94461301e7a6ede06

**Reply by treykeown** [reply]
My use case is a custom de/serializer. I only support a few specific types - it's not able to build subclasses. For now I work around this with runtime assertion:
```py
from typing import get_args

AnySerializable: TypeAlias = Foo | Bar
serializable_types = get_args(AnySerializable)

def serialize(item: AnySerializable) -> bytes:
    assert type(item) in serializable_types
    ...
```

**Reply by erictraut** [reply, marked as answer]
There's no way to specify an "exact type" in any typing construct. From a type theory standpoint, any subclass of `Foo` is also a `Foo`.

One potential solution is to mark `Foo` as `@final` which would prevent any subclasses from being defined.
